packages/inertia/src/inertia.js

```diff
@@ -161,6 +214,9 @@ export default {
         if (only.length && response.data.component === this.page.component) {
           response.data.props = { ...this.page.props, ...response.data.props }
         }
+        if (url.hash && this.urlWithoutHash(url) === this.normalizeUrl(response.data.url).href) {
```

Comment:
Maybe we can extract this to a small helper method so that it's not as verbose? Maybe something like `this.isSamePageHashVisit`?

Reply:
Hmm, I'm realizing that we actually need to check here if the response URL has a hash, otherwise we'll end up with multiple hashes. This is actually what we need:

```js
const responseUrl = this.normalizeUrl(response.data.url)
if (url.hash && !responseUrl.hash && this.urlWithoutHash(url) === responseUrl.href) {
  responseUrl.hash = url.hash
  response.data.url = responseUrl.href
}
```